src/nnet3/nnet-attention-component.cc

```diff
@@ -0,0 +1,761 @@
+      num_right_inputs_required_ > num_right_inputs_ ||
+      time_stride_ <= 0)
+    KALDI_ERR << "Config line contains invalid values: "
+              << cfl->WholeLine();
+  stats_count_ = 0.0;
+  Check();
+}
+
+
+
+virtual void*
+RestrictedAttentionComponent::Propagate(const ComponentPrecomputedIndexes *indexes_in,
+                                        const CuMatrixBase<BaseFloat> &in,
+                                        CuMatrixBase<BaseFloat> *out) const {
+  const PrecomputedIndexes *indexes = dynamic_cast<const PrecomputedIndexes*>(
+      indexes_in);
+  KALDI_ASSERT(indexes != NULL &&
+               in.NumRows() == indexes->io.num_t_in * indexes->io.num_images &&
+               out->NumRows() == indexes->io.num_t_out * indexes->io.num_images);
+
+  // mostly a wrapper for PropagateOneHead.
+
+}
+
+void RestrictedAttentionComponent::PropagateOneHead(
+    const time_height_convolution::ConvolutionComputationIo &io,
+    const CuMatrixBase<BaseFloat> &in,
+    CuMatrixBase<BaseFloat> *c,
+    CuMatrixBase<BaseFloat> *out) const {
+  int32 query_dim = key_dim_ + context_dim_,
+      full_value_dim = key_dim_ + (output_context_ ? context_dim_ : 0);
+  KALDI_ASSERT(in.NumRows() == io.num_images * io.num_t_in &&
+               out->NumRows() == io.num_images * io.num_t_out &&
+               out->NumCols() == full_value_dim &&
+               in.NumCols() == (key_dim_ + value_dim_ + query_dim_) &&
+               io.t_step_in == io.t_step_out &&
+               (io.start_t_out - io.start_t_in) % io.t_step_in == 0 &
+               (io.num_t_out - io.num_t_in) % io.t_step_in == 0);
+
+  // 'steps_left_context' is the number of time-steps the input has on the left
+  // that don't appear in the output.
+  int32 steps_left_context = (io.start_t_out - io.start_t_in) / io.t_step_in,
+      rows_left_context = steps_left_context * io.num_images;
+  KALDI_ASSERT(rows_left_context >= 0);
+
+  // 'queries' contains the queries.  We don't use all rows of the input
+  // queries; only the rows that correspond to the time-indexes at the
+  // output, i.e. we exclude the left-context and right-context.
+  // 'out'; the remaining rows of 'in' that we didn't select correspond to left
+  // and right temporal context.
+  CuSubMatrix<BaseFloat> queries(in, rows_left_context, out->NumRows(),
+                                 key_dim_ + value_dim_, query_dim_);
+  // 'keys' contains the keys; note, these are not extended with
+  // context information; that happens further in.
+  CuSubMatrix<BaseFloat> keys(in, 0, in.NumRows(), 0, key_dim_);
+
+  // 'values' contains the values which we will interpolate.
+  // these don't contain the context information; that will be added
+  // later if output_context_ == true.
+  CuSubMatrix<BaseFloat> values(in, 0, in.NumRows(), key_dim_, value_dim_);
+
+
+  AttentionForward(key_scale_, keys, queries, values, c, out);
+}
+
+
+
+void TimeHeightConvolutionComponent::Check() {
```

Comment:
Is `TimeHeightConvolutionComponent` supposed to be here? Almost all the functions ralating `TimeHeightConvolutionComponent` are duplicated here.